shells/android/java/arcs/crdt/parcelables/ParcelableCrdtEntity.kt

```diff
@@ -0,0 +1,292 @@
+                    val data = requireNotNull(
+                        parcel.readTypedObject(ParcelableCrdtSingleton.Data.CREATOR)
+                    ).actual
+                    singletons[field] = CrdtSingleton.createWithData(
+                        data as CrdtSingleton.Data<CrdtEntity.Reference>
+                    )
+                }
+
+                val collections = mutableMapOf<FieldName, CrdtSet<CrdtEntity.Reference>>()
+                val numCollections = parcel.readInt()
+                repeat(numCollections) {
+                    val field = requireNotNull(parcel.readString())
+                    val data = requireNotNull(
+                        parcel.readTypedObject(ParcelableCrdtSingleton.Data.CREATOR)
+                    ).actual
+                    collections[field] =
+                        CrdtSet.createWithData(data as CrdtSet.Data<CrdtEntity.Reference>)
+                }
+
+                return Data(CrdtEntity.Data(versionMap.actual, singletons, collections))
+            }
+
+            override fun newArray(size: Int): Array<Data?> = arrayOfNulls(size)
+        }
+    }
+
+    /**
+     * Parcelable variants of [CrdtEntity.Operation].
+     *
+     * This class is implemented such that it serves as a multiplexed parcelable for its subclasses.
+     * We write the ordinal value of [OpType] first, before parceling the contents of the subclass.
+     * The [OpType] is used to figure out the correct subtype when deserialising.
+     */
+    sealed class Operation(
+        private val opType: OpType
+    ) : ParcelableCrdtOperation<CrdtEntity.Operation> {
+
+        override fun writeToParcel(parcel: Parcel, flags: Int) {
+            // Write the opType so we can multiplex during createFromParcel.
+            parcel.writeInt(opType.ordinal)
+        }
+
+        /** Parcelable variant of [CrdtEntity.Operation.SetSingleton]. */
+        data class SetSingleton(
+            override val actual: CrdtEntity.Operation.SetSingleton
+        ) : Operation(OpType.SetSingleton) {
+            override fun describeContents(): Int = 0
+
+            override fun writeToParcel(parcel: Parcel, flags: Int) {
+                super.writeToParcel(parcel, flags)
+                parcel.writeTypedObject(actual.clock.toParcelable(), flags)
+                parcel.writeString(actual.actor)
+                parcel.writeString(actual.field)
+                parcel.writeTypedObject(actual.value.toParcelable(), flags)
+            }
+
+            companion object CREATOR : Parcelable.Creator<SetSingleton> {
+                override fun createFromParcel(parcel: Parcel): SetSingleton {
+                    val clock =
+                        requireNotNull(parcel.readTypedObject(ParcelableVersionMap.CREATOR)) {
+                            "VersionMap not found in parcel when reading " +
+                                "ParcelableCrdtEntity.Operation.SetSingleton"
+                        }.actual
```

Comment:
Consider using the `parcel.readVersionMap()` extension function (if it exists), and moving the `.actual` call within the `requireNotNull`

Reply:
Done